src/alexandria3k/data_sources/uspto.py

```diff
@@ -18,71 +18,581 @@
 #
 """Patent grant bibliographic (front page) text data (JAN 1976 - present)"""
 
+import zipfile
+import os
+import re
+
+from alexandria3k.data_source import (
+    CONTAINER_INDEX,
+    ROWID_INDEX,
+    DataSource,
+    ElementsCursor,
+    StreamingCachedContainerTable,
+)
+
+from alexandria3k.xml import getter, agetter, all_getter
+from alexandria3k.file_xml_cache import get_file_cache
 from alexandria3k.db_schema import ColumnMeta, TableMeta
 
+
 # Bulk data can be found here. https://bulkdata.uspto.gov
 # Patent Grant Bibliographic (Front Page) Text Data (JAN 1976 - PRESENT)
-DEFAULT_SOURCE = None
+DEFAULT_SOURCE = (
+    "path/to/tests/data/April 2023 Patent Grant Bibliographic Data"
+)
 
+# Delimiter for extracting concatenated XML files.
+XML_DELIMITER = '<?xml version="1.0" encoding="UTF-8"?>'
 
 # Dataset Description — Patent grant full-text data (no images)
 # JAN 1976 — present Automated Patent System (APS)
 # Contains the full text of each patent grant issued weekly
 # Tuesdays from January 1, 1976, to present excludes images/drawings and reexaminations.
 # https://developer.uspto.gov/product/patent-grant-bibliographic-dataxml
 
+
+class ErrorElement:
+    """A class used for representing error elements"""
+
+    def find(self, _path):
+        """Placeholder to the actual XML tree element."""
+        return None
+
+    def findall(self, _path):
+        """Placeholder to the actual XML tree element."""
+        return None
+
+
+class ZipFiles:
+    """The source of the compressed XML data files"""
+
+    def __init__(self, directory, sample_container):
+        # Collect the names of all available data files
+        self.file_path = []
+        self.unique_patent_xml_files = []
+        self.patent_file_name = []
+        self.sample = sample_container
+        pattern = r'file="([^"]+)"'
+
+        # Read through the directory
+        for file_name in os.listdir(directory):
+            path = os.path.join(directory, file_name)
+            # print(path)
+            if not os.path.isfile(path):
+                continue
+            if not sample_container(path):
+                continue
+
+            # Collect the names of all available XML zip data files
+            self.file_path.append(path)
+
+        # pylint: disable-next=consider-using-with
+        for uspto_record in self.file_path:
+            # perf.log("Reading USPTO zip files...")
+            # Open every US patent XML zip file
+            with zipfile.ZipFile(uspto_record, "r") as zip_ref:
+                # Pick only the XML file inside zip
+                xml_file = [
+                    file
+                    for file in zip_ref.namelist()
+                    if file.endswith(".xml")
+                ]
+                (self.file_name,) = xml_file
+
+                # Get chunks of concatenated XML files
+                with zip_ref.open(self.file_name, "r") as uspto_weekly_file:
```

Comment:
So here you are decompressing all zip files when the object is constructed?
This should happen dynamically and on-demand, so that object construction is fast.
Please try your code on the entire USPTO data set to set that it behaves reasonably, e.g. it responds fast in a `LIMIT 1` query.
Note: It might be OK to simply obtain the file contents, so iterating through the zip files and their contents, without calling `open`.